### PR TITLE
2022 Update to The Annotated Transformer

### PR DESCRIPTION
**Pull request by austinvhuang**
# Goals

1. Get the notebook to run on a recent version of pytorch and supporting libraries (torchtext, altair).
2. Simplify and reduce implementation complexity where new/standard functionality accomplishes the same thing.
3. Reflect current standard practices (remove deprecated abstractions, replace DP with DDP).
5. Try to improve ease of comprehension where possible. Keep the core pedagogical content roughly the same as the original.

# Change Notes

- Runs on a recent version of pytorch (0.11) 
- Use jupytext to pair notebook with python file
- Add github actions to check for package installation success and flake standards compliance
- Added more detailed build instructions to readme
- All plots use altair (originally was a mix of matplotlib and seaborn)
- Eliminate use of deprecated Variable type.
- Flatten/reduce abstractions in favor of off-the-shelf pytorch - replaced custom NoamOpt wrapper with LambdaLR for custom learning rate schedule.
- Flatten/reduce abstractions in favor of off-the-shelf pytorch - Removed custom MyIterator wrapper, used standard collate_fn parameter to DataLoader to prep batches as tensors
- Refactored data loading to be compatible with torchtext 0.12, removed deprecated abstractions like Field 
- Automated loading in code of spacy artifacts rather than require manual installation
- Updated and fixed token indices (e.g. start/end of sentence).
- Reduce use of global variables and wrap all cells in functions (also necessary for DDP because when processes are spawned, it imports the worker function from the python script and can't be executing code on import)
- Replace conditional branching code on None with passing in a no-op  (DummyOptimizer)
- Replaced data parallel (DP) and custom multi-gpu loss with distributed data parallel (DDP) training
- Show multiple example outputs instead of just one
- Attention maps are interactive with tooltips when mousing over each cell
- Split organization into 3 sections: 1) architecture 2) model training setup 3) real-world translation task

# Feedback Requests

- General feedback on above changes, overall state as well as any specific changes needed.

- Some compromises in standard notebook practices were made so that it could be compatible with DDP - specifically wrapping everything in functions and putting a test condition around executed script code. Is the notebook comprehensible in spite of these compromises? Is someone seeing it for the first time going to be confused?

- Haven't added collaboration team attribution for this update - let me know if you have any preference where to put that @srush .

### CODE REVIEW

**Comment by srush** [comment]
This looks great to me. I think ideally section 3 would have some more explanations, but section 1 and section 2 looks like major improvements. Let's merge and work on getting a version up.
